File: intermediate/18_steamroller.js
//STEAMROLLER
//Flatten a nested array. You must account for varying levels of nesting.

const flattenManyToOne = array =>
  array.reduce((acc, cur) => acc.concat(cur), []);

const flattenOneToOne = item => {
  while (Array.isArray(item)) {
    item = item[0];
  }
  return item;
};

const steamrollArray = array => {
  return flattenManyToOne(array).map(flattenOneToOne);
};

//Tests
steamrollArray([[['a']], [['b']]]); //should return ["a", "b"].
steamrollArray([1, [2], [3, [[4]]]]); //should return [1, 2, 3, 4].
steamrollArray([1, [], [3, [[4]]]]); //should return [1, 3, 4].
steamrollArray([1, {}, [3, [[4]]]]); //should return [1, {}, 3, 4].

File: intermediate-python/18_steamroller.py
#STEAMROLLER
#Flatten a nested array. You must account for varying levels of nesting.

from functools import reduce

def flat(acc, cur):
  if type(cur) == type([]):
    acc.extend(cur)
  else:
    acc.append(cur)
  return acc

def flattenManyToOne(array):
  return reduce(flat, array, [])

def flattenOneToOne(item):
  while type(item) == type([]):
    item = item[0]
  return item

# OR using recursion:
def flattenOneToOne(item):
  if type(item) != type([]):
    return item
  else:
    return flattenOneToOne(item[0])

def steamrollArray(array):
  return list(map(flattenOneToOne, flattenManyToOne(array)))


#Tests
print (steamrollArray([['a']])) #should return ['a']
print (steamrollArray([[['a']], [['b']]])) #should return ["a", "b"].
print (steamrollArray([1, [2], [3, [[4]]]])) #should return [1, 2, 3, 4].
print (steamrollArray([1, [], [3, [[4]]]])) #should return [1, 3, 4].
print (steamrollArray([1, {}, [3, [[4]]]])) #should return [1, {}, 3, 4].
